Model/Api/DiscountCodeValidation.php

```diff
@@ -288,130 +295,213 @@ public function validate() {
                 );
             }
 
-            // Check date validity if "To" date is set for the rule
-            $date = $rule->getToDate();
-            if ($date && date('Y-m-d', strtotime($date)) < date('Y-m-d')) {
-                return $this->sendErrorResponse(
-                    self::ERR_CODE_EXPIRED,
-                    sprintf('The code [%s] has expired.', $couponCode),
-                    422,
-                    $immutableQuote
-                );
+            if ($coupon && $coupon->getCouponId()) {
+                $result = $this->applyingCouponCode($couponCode, $coupon, $immutableQuote, $parentQuote);
+            } else if ($giftCard && $giftCard->getId()) {
+                $result = $this->applyingGiftCardCode($couponCode, $giftCard, $immutableQuote, $parentQuote);
+            } else {
+                throw new Exception(__('Something happened with current code.'));
             }
 
-            // Check date validity if "From" date is set for the rule
-            $date = $rule->getFromDate();
-            if ($date && date('Y-m-d', strtotime($date)) > date('Y-m-d')) {
-                $desc = 'Code available from ' . $this->timezone->formatDate(
-                        new \DateTime($rule->getFromDate()),
-                        \IntlDateFormatter::MEDIUM
-                    );
-                return $this->sendErrorResponse(
-                    self::ERR_CODE_NOT_AVAILABLE,
-                    $desc,
-                    422,
-                    $immutableQuote
-                );
-            }
+            return $this->sendSuccessResponse($result, $immutableQuote);
```

Comment:
applyingCouponCode method may call sendErrorResponse method internally. I guess in such case we shouldn't call sendSuccessResponse?

Reply:
Yes sure, i will think and refactor it.

Reply:
I've added check above